### PR TITLE
Make backtrace as a cargo feature

### LINKED ISSUE #7522: Performance Regression: Backtraces in errors slow down planning time (Expensive backtraces)
DataFusion creates loads of errors even on the happy path. However as of #7434, we now gather a backtrace for each error. This is rather expensive. Here is a profile dump from a prod workload:

![Screenshot from 2023-09-11 18-35-55](https://github.com/apache/arrow-datafusion/assets/1529400/c6541616-18dd-44a3-b0ce-2f82500b1b9f)

In the said workload, there is a LOT of going for bookkeeping (that's why I had the profiler running in the first place) but the backtraces alone make up for 30% of the time. The place looks like this:

```text
logical:
Projection: ...
  TableScan: ...                                                                                                                                                                                                                                                                                                                                                                                               

physical:
ProjectionExec: ..
  CoalesceBatchesExec: target_batch_size=8192
    FilterExec: ...
      ParquetExec: ...
```

(had to remove a good amount of details due to data protection, but the filters / predicates are rather simple)

I think there are two paths forward:

- do NOT generate backtraces for errors
- do NOT use errors for the happy path but rather `Option` or some other enum

### PR DESCRIPTION
**Pull request by comphead**
## Which issue does this PR close?

Closes #7522.

## Rationale for this change
In #7434 the backtrace was introduced, however its expensive to build and should be optional. Introducing a cargo feature for backtrace

## Are these changes tested?
yes

### CODE REVIEW

**Comment by comphead** [comment]
@alamb @crepererum

**Review by crepererum — COMMENTED** [review]
This adds a build config, not a cargo feature. A cargo feature (which I think should be preferred because build configs are an even bigger mess) works like this:

1. add feature to `Cargo.toml`
2. change the cfg guard to `#[cfg(feature = "backtrace")]`

Also see <https://doc.rust-lang.org/cargo/reference/features.html>.

On top of that, I think this needs some docs, at least for `get_back_trace` (you probably want to use a single method so you can easily reuse the docstring and either use the cfg-guards within that method or create a new private fn called `get_back_trace_impl`).

File: datafusion/common/src/error.rs [hunk 1]
```diff
@@ -412,20 +415,34 @@ impl DataFusionError {
...
-            .split("\n\nbacktrace: ")
+            .split(Self::BACK_TRACE_SEP)
             .collect::<Vec<&str>>()
             .first()
             .unwrap_or(&"")
             .to_string()
     }
 
+    /// To enable optional rust backtrace in DataFusion:
+    /// - [`Setup Env Variables`]<https://doc.rust-lang.org/std/backtrace/index.html#environment-variables>
```
Comment (alamb) [inline datafusion/common/src/error.rs:426]: 👍

File: datafusion/common/Cargo.toml [hunk 2]
```diff
@@ -34,6 +34,7 @@ path = "src/lib.rs"
 
 [features]
 avro = ["apache-avro"]
```
Comment (alamb) [inline datafusion/common/Cargo.toml:36]: As written, to be enabled, a user would have to use `datafusion_common` directly (they couldn't activate it through `datafusion`)

To enable it via datafusion we also need to add it to core/Cargo.toml, like this:

https://github.com/apache/arrow-datafusion/blob/cde74016e930ffd9c55eed403b84bcd026f38d0f/datafusion/core/Cargo.toml#L44

File: datafusion/common/src/error.rs [hunk 3]
```diff
@@ -522,6 +539,41 @@ mod test {
         assert_eq!(res.strip_backtrace(), "Arrow error: Schema error: bar");
     }
 
+    // RUST_BACKTRACE=1 cargo test --features backtrace --package datafusion-common --lib -- error::test::test_backtrace
+    #[cfg(feature = "backtrace")]
+    #[test]
+    #[allow(clippy::unnecessary_literal_unwrap)]
+    fn test_enabled_backtrace() {
```
Comment (alamb) [inline datafusion/common/src/error.rs:546]: I wonder if we should also add a CI test, perhaps like https://github.com/apache/arrow-datafusion/blob/cde74016e930ffd9c55eed403b84bcd026f38d0f/.github/workflows/rust.yml#L99

**Review by alamb — COMMENTED** [review]
Thank you @comphead  -- I think this is looking very close

File: datafusion/common/Cargo.toml (hunk 2 above)
Comment (comphead) [reply datafusion/common/Cargo.toml:36]: Thanks I was little bit struggling on how it make on the datafusion level

**Comment by comphead** [comment]
2 tests failed because the configuration is different between architectures for 
amd:
```
 env:
    RUSTFLAGS: -C debuginfo=1
    CARGO_INCREMENTAL: 0
    RUST_BACKTRACE: 1
```

mac,win
```
  env:
    RUSTFLAGS: -C debuginfo=0
``` 

Lack of BACKTRACE=1 on mac,win arch expectedly lead the test to fail. @alamb would you mind to help where this set up?

**Review by alamb — APPROVED** [review]
> Lack of BACKTRACE=1 on mac,win arch expectedly lead the test to fail. @alamb would you mind to help where this set up?

It looks like you have solved this already @comphead  -- let me know if there is something else you are looking for

Otherwise I think this PR is looking ready to go. Thank you 🙏

**Pull request merged**